mapswipe_workers/mapswipe_workers/firebase_to_postgres/update_data.py

```diff
@@ -1,129 +1,233 @@
 """Update users and project information from Firebase in Postgres."""
 
+import asyncio
+import concurrent.futures
+import csv
 import datetime as dt
+import io
+from typing import List, Optional
 
 from mapswipe_workers import auth
-from mapswipe_workers.definitions import logger, sentry
+from mapswipe_workers.definitions import logger
 
 
-def get_last_updated_timestamp() -> str:
-    """Get the timestamp of the latest created user in Postgres."""
-    pg_db = auth.postgresDB()
-    query = """
-        SELECT created
-        FROM users
-        WHERE created IS NOT NULL
-        ORDER BY created DESC
-        LIMIT 1
-        """
-    last_updated = pg_db.retr_query(query)
+def get_user_attribute_from_firebase(user_ids: List[str], attribute: str):
```

Comment:
I am not sure what the goal is to use asyncio here.

In most use cases if you want to execute async code it should be enough to use [asyncio.run](https://docs.python.org/3/library/asyncio-task.html#asyncio.run) as executor. From the docs:
> Execute the coroutine coro and return the result.
> [...]
> This function always creates a new event loop and closes it at the end. It should be used as a main entry point for asyncio programs, and should ideally only be called once."

Probably asyncio.run is not used here because of the usage of `ThreadPoolExecutor`. I think that the usage of `event_loop.run_in_executer` + `ThreadPoolExecutor` is a way to run blocking code in a async function. See also [Python docs](https://docs.python.org/3/library/asyncio-dev.html#running-blocking-code) and [this Stackoverflow post](https://stackoverflow.com/questions/59199218/when-you-need-to-use-asyncio-and-threadpoolexecutor-do-you-need-to-call-loop-cl) .

Since the Firebase SDK has not been implementing the async/await pattern or async in general I would recommend to use parallel execution using the threading module of Python. Or was there a reason not to use threading?

Another possibility would be to use the REST-API of Firebase with an async http library as @Hagellach37 proposed in the comments already.

After reading the in the docstring mentioned blog post the implemented code is in fact to run IO-Blocking code embedded in an async code base using threading. This is only useful if the rest of the code is asynchronous. Which is not the case here. The efficiency here will be the same using threading but the code will be much more simpler.

Reply:
Hey @Matthias-Schaub ,
thanks for you suggestion here. I looked into it a bit more detailed and you are right, that it's much easier to use just the threading module here.